buster/retriever/deeplake.py

```diff
@@ -9,17 +11,67 @@
 logging.basicConfig(level=logging.INFO)
 
 
+def extract_metadata(x, columns):
+    """Returned metadata from deeplake is in a nested dict, extract it so that each attribute has its own column."""
+    for col in columns:
+        x[col] = x.metadata[col]
+    return x
+
+
+def data_dict_to_df(data: dict):
+    # rename 'score' to 'similarity'
+    data["similarity"] = data.pop("score")
+    data["content"] = data.pop("text")
+
+    matched_documents = pd.DataFrame(data)
+
+    if len(matched_documents) == 0:
+        logger.info("No matches found...")
+        return pd.DataFrame()
+
+    matched_documents = matched_documents.apply(extract_metadata, columns=["source", "title", "url"], axis=1)
+    matched_documents = matched_documents.drop(columns="metadata")
+
+    return matched_documents
+
+
+def build_tql_query(embedding, sources=None, top_k: int = 3):
+    # Initialize the where_clause to an empty string.
+    where_clause = ""
+
+    embedding_string = ",".join([str(item) for item in embedding])
+
+    # If sources is provided and it's not empty, build the where clause.
+    if sources:
+        conditions = [f"contains(metadata['source'], '{source}')" for source in sources]
+        where_clause = "where " + " or ".join(conditions)
+
+    # Construct the entire query
+    query = f"""
+select * from (
```

Comment:
I don't understand why you need the `select * from`

Reply:
not sure, it's deeplake's TQL which im not too familiar with

Reply:
This one in particular I wrote myself based on examples from their docs, https://docs.deeplake.ai/en/latest/Tensor-Query-Language.html

Reply:
then try this one:
`select embedding, text, metadata, cosine_similarity(embedding, ARRAY[{embedding_string}]) as score {where_clause} order by score desc limit {top_k}`

Reply:
it fails :(

Reply:
alright then, let's forget it. It's a very weird looking query, but I don't want to look into their doc to make it nicer.